include/revng/Pipeline/InputOutputContract.h

```diff
@@ -0,0 +1,201 @@
+/// into functions with quantifiers(i, *).
+///
+/// A contract from functions(granularity 2) to root(granularity 1), requires
+/// that the pipe it is attached to will convert all functions with quantifiers
+/// (i, *) into roots with quantifiers(i). Furthermore it is guaranteed that
+/// only functions in the form (i, *) are presents, in other words that either
+/// all or none functions are forwarded to the pipe.
+///
+/// A pipe marked with a single contract must transform all and only items from
+/// the input container that have Targets with the same kind as if SourceKind
+/// (or a derived kind if Exactness is not Exact)
+///
+/// The pipe must furthermore esures that for each item in the input container
+/// that matches the requirements another target is created in the target
+/// container (is the target container exists), such that
+///
+/// If the output kind has a greater granularity than the inputkind then for all
+/// Targets in input with source kind and quantifiers (i1, ..., in) then a
+/// target with kind output and quantifiers (i1, ..., in, *) exists.
+///
+/// if the output has a smaller granularity g1 than the input granularity g2
+/// then for the pipe must operate on all targets with Kind source and
+/// quanfifiers (i1, ..., i_g2) and yield a a target for each with Kind target
+/// and quantifiers(i1, ..., i_g1), it will be ensured that when the pipe
+/// executed each target in the input container that must be transformed will be
+/// in the form (i1, ..., i_g2) where each quantifier i_x such that g1 <= x <=
+/// g2 is *.
+///
+///
+class InputOutputContract {
+private:
+  const Kind *Source;
+  Exactness InputContract;
+  const Kind *TargetKind;
+  size_t PipeArgumentSourceIndex;
+  size_t PipeArgumentTargetIndex;
+  bool PreservedInput;
+
+public:
+  constexpr InputOutputContract(const Kind &Source,
+                                Exactness InputContract,
+                                size_t PipeArgumentSourceIndex,
+                                const Kind &Target,
+                                size_t PipeArgumentTargetIndex = 0,
+                                bool PreservedInput = false) :
+    Source(&Source),
+    InputContract(InputContract),
+    TargetKind(&Target),
+    PipeArgumentSourceIndex(PipeArgumentSourceIndex),
+    PipeArgumentTargetIndex(PipeArgumentTargetIndex),
+    PreservedInput(PreservedInput) {}
+
+  constexpr InputOutputContract(const Kind &Target,
+                                size_t PipeArgumentTargetIndex = 0) :
+    Source(nullptr),
+    InputContract(Exactness::Exact),
+    TargetKind(&Target),
+    PipeArgumentSourceIndex(0),
+    PipeArgumentTargetIndex(PipeArgumentTargetIndex),
+    PreservedInput(false) {}
+
+  constexpr InputOutputContract(const Kind &Source,
+                                Exactness InputContract,
+                                size_t PipeArgumentSourceIndex = 0,
+                                bool PreservedInput = false) :
```

Comment:
`bool` arguments can be quite obsure on call sites.
What do you think about having an `enum` with just two entries?